scenario/state.py

```diff
@@ -643,24 +644,35 @@ def _generate_new_change_id():
 
 
 @dataclasses.dataclass(frozen=True)
-class ExecOutput(_max_posargs(0)):
+class Exec(_max_posargs(1)):
     """Mock data for simulated :meth:`ops.Container.exec` calls."""
 
+    command_prefix: Sequence[str]
     return_code: int = 0
     """The return code of the process (0 is success)."""
     stdout: str = ""
     """Any content written to stdout by the process."""
     stderr: str = ""
     """Any content written to stderr by the process."""
+    stdin: Optional[str] = None
+    """Any content written to stdin by the charm."""
 
     # change ID: used internally to keep track of mocked processes
     _change_id: int = dataclasses.field(default_factory=_generate_new_change_id)
 
+    def __post_init__(self):
+        # The command prefix can be any sequence type, and a list is tidier to
```

Comment:
why are we not making the command_prefix a string? we can `shlex.split` it later to do the prefix matching thing.
Or are you trying to keep the syntax Popen-like?

Reply:
It's a list when you do `self.unit.get_container("foo").exec()` so I feel it ought to match (which in turn was probably inspired by Popen).

When I've looked over how people use this in Harness, I think there's a bunch that just do a copy and paste from the charm code to get the full command and would want to support that, as well as people that do the plain `[]` knowing that there's only one process, or do the slightly nicer version where it's just the main executable.

For what it's worth, it's also a more minimal change from current Scenario, where it's a tuple as the dictionary key.

Reply:
yeah I get the consistency argument. Fine with the way it is.